src/corecel/sys/ScopedProfiling.hh

```diff
@@ -0,0 +1,76 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/sys/ScopedProfiling.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <string>
+
+#include "celeritas_config.h"
+
+namespace celeritas
+{
+
+//---------------------------------------------------------------------------//
+/*!
+ * Input arguments for the nvtx implementation.
+ */
+struct ScopedProfilingInput
+{
+    std::string name;  //!< Name of the range
+    uint32_t color{0xFF00FF00};  //!< ARGB
+    int32_t payload{0};  //!< User data
+    uint32_t category{0};  //!< Category, used to group ranges together
+
+    ScopedProfilingInput(std::string const& name) : name{name} {}
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * RAII class for scoped profiling.
+ *
+ * Implementations should support multithreaded context where each thread have
+ * one or more alive instance of this class.
+ *
+ * This is useful for wrapping specific code fragment in a range for profiling,
+ * e.g. ignoring of VecGeom instantiation kernels, profiling a specific action
+ * or loop on the CPU.
+ * TODO: Template ScopedProfiling over profiling backend if we need to add a
+ * new one
+ */
+class ScopedProfiling
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Input = ScopedProfilingInput;
+    //!@}
+
+  public:
+    // Activate profiling with options
+    explicit ScopedProfiling(Input input);
+    // Activate profiling
+    explicit ScopedProfiling(std::string const& name);
+    // Deactivate profiling
+    ~ScopedProfiling();
+    // Can't outlive the scope it was created in to correctly match each push /
+    // pop. Must strictly follows RAII semantics (automatic storage duration)
+    void* operator new(std::size_t) = delete;
+    ScopedProfiling(ScopedProfiling const&) = delete;
+    ScopedProfiling& operator=(ScopedProfiling const&) = delete;
+    ScopedProfiling(ScopedProfiling&&) = delete;
+    ScopedProfiling& operator=(ScopedProfiling&&) = delete;
```

Comment:
Remove the `operator new` specialization, and the `&&` instances are automatically deleted if `&` is deleted so you can omit those.

Reply:
I can see allowing `operator new` even though using it is asking for trouble. 

Regarding the move operations, should we be explicit and follow [C.21](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all)?

Reply:
Good call. It looks like `ScopedMem`, `Stream`, and `ScopedSignalHandler` all need that as well, and a few "scoped" classes are entirely missing the deleted copy/assign:
- src/celeritas/ext/ScopedGeantExceptionHandler.hh
- src/celeritas/ext/ScopedGeantLogger.hh
- src/celeritas/ext/ScopedRootErrorHandler.hh
- src/corecel/io/ScopedStreamFormat.hh
- src/corecel/io/ScopedStreamRedirect.hh
- src/corecel/io/ScopedTimeAndRedirect.hh
- src/corecel/io/ScopedTimeLog.hh
- src/corecel/sys/ScopedMpiInit.hh

Should those updates be a separate PR too? If you haven't pushed the restoration of `..&&) = delete`  I think we should do them all in a separate best-practices merge.

Reply:
sure, we can do that in a separate PR.